test/e2e/utils.go

```diff
@@ -728,3 +741,120 @@ func getEnvAndExpectNoError(key string) string {
 	utilruntime.Must(err)
 	return val
 }
+
+// newTestHelperJob returns the K8s Job for given commands to be executed inside k8s cluster.
+// This test helper job can be used to validate test cases from inside the k8s cluster by executing the bash scripts.
+func newTestHelperJob(jobName string, podSpec *corev1.PodSpec) *batchv1.Job {
+	return &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      jobName,
+			Namespace: namespace,
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: *podSpec,
+			},
+			BackoffLimit: pointer.Int32(0),
+		},
+	}
+}
+
+// etcdZeroDownTimeValidatorJob returns k8s job which ensures
+// Etcd cluster zero down time by continuously checking etcd cluster health.
+// This job fails once health check fails and associated pod results in error status.
+func etcdZeroDownTimeValidatorJob(etcdSvc, testName string, tls *v1alpha1.TLSConfig) *batchv1.Job {
+	return newTestHelperJob(
+		"etcd-zero-down-time-validator-"+testName,
+		&corev1.PodSpec{
+			Volumes: []v1.Volume{
+				{
+					Name: "client-url-ca-etcd",
+					VolumeSource: v1.VolumeSource{
+						Secret: &v1.SecretVolumeSource{
+							SecretName:  tls.TLSCASecretRef.Name,
+							DefaultMode: pointer.Int32(420),
+						},
+					},
+				},
+				{
+					Name: "client-url-etcd-server-tls",
+					VolumeSource: v1.VolumeSource{
+						Secret: &v1.SecretVolumeSource{
+							SecretName:  tls.ClientTLSSecretRef.Name,
+							DefaultMode: pointer.Int32(420),
+						},
+					},
+				},
+				{
+					Name: "client-url-etcd-client-tls",
+					VolumeSource: v1.VolumeSource{
+						Secret: &v1.SecretVolumeSource{
+							SecretName:  tls.ServerTLSSecretRef.Name,
+							DefaultMode: pointer.Int32(420),
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:    "etcd-zero-down-time-validator-" + testName,
+					Image:   "alpine/curl",
+					Command: []string{"/bin/sh"},
+					//To avoid flakiness, consider downtime when curl fails consecutively back-to-back.
+					Args: []string{"-ec",
+						"echo '" +
+							"failed=0 ; threshold=2 ; " +
+							"while [ $failed -lt $threshold ] ; do  " +
+							"$(curl --cacert /var/etcd/ssl/client/ca/ca.crt --cert /var/etcd/ssl/client/client/tls.crt --key /var/etcd/ssl/client/client/tls.key https://" + etcdSvc + ":2379/health -s -f  -o /dev/null ); " +
+							"if [ $? -gt 0 ] ; then let failed++; echo \"etcd is unhealthy and retrying\"; continue;  fi ; " +
+							"echo \"etcd is healthy\";  touch /tmp/healthy; let failed=0; " +
+							"sleep 1; done;  echo \"etcd is unhealthy\"; exit 1;" +
+							"' > test.sh && sh test.sh",
+					},
+					ReadinessProbe: &v1.Probe{
+						InitialDelaySeconds: int32(5),
+						FailureThreshold:    int32(1),
+						PeriodSeconds:       int32(1),
+						SuccessThreshold:    int32(3),
```

Comment:
why `successThreshold` is set to `3` ? Is successThreshold:`1` not good enough ?

Reply:
After two Curl fails continuously, we exit with code 1.
successThreshold is `3` pod to be ready, at least two curls to be succeeded out of 3 curls, this will reduce  fake signal of cluster health before starting next test case